### DISCUSSION TITLE
Achieve loose coupling for Interceptors using attributes

### DISCUSSION THREAD

**Discussion by git4rputuval** [quarkusio/quarkus#38050] [Q&A]
The purpose is to use Interceptor for my service and add an attribute (`serviceName`) to the custom ( interceptor) annotation, where the service name would be passed by every respective service annotated by this interceptor. 

<img width="960" alt="InterceptorInterface" src="https://github.com/quarkusio/quarkus/assets/96055225/5885b197-dc36-4021-ae47-da71dea9ce47">

However, this attribute name needs to be passed in the interceptor implementation and the service classes. The attribute can be passed in the service class as shown in the snapshot below.

<img width="960" alt="Service" src="https://github.com/quarkusio/quarkus/assets/96055225/0ad2736b-c2ad-4b52-9638-2a12dbe46324">

But it cannot be passed in the interceptor implementation due to which we have a compilation problem as shown in snapshot below. 
<img width="960" alt="implementationClassOfInterceptor" src="https://github.com/quarkusio/quarkus/assets/96055225/c9d05140-4087-481d-9f86-01ff5a77c7bc">

We do not want to hardcode or re-enter the service name in the Interceptor Implementation class since it defeats the purpose of an interceptor being used across the system. Instead it should extract the attribute value from the service annotated by the Interceptor and marked with the `serviceName` attribute. 

How do we resolve the compilation issue? 

please find the attached zip to run the code n check for error or workaround for a solution

[InterceptorIssue.zip](https://github.com/quarkusio/quarkus/files/13839264/InterceptorIssue.zip)

**Comment by Postremus** [comment]
You can use a `@NonBinding` on your serviceName member.

For example:
```
@InterceptorBinding
@Target({
        TYPE, METHOD
})
@Retention(RUNTIME)
@Inherited
public @interface JobMetadataLock {

    @Nonbinding
    JobMetadataType type();

}
```

And my interceptor looks like this:

```
@JobMetadataLock(type = JobMetadataType.XX)
@Interceptor
@Priority(APPLICATION)
public class JobMetadataLockInterceptor {
@AroundInvoke
    public Object invoke(InvocationContext ctx) throws Exception {
..
}
}
```

The NonBinding member (type in my case) still needs to be specified on the interceptor, since it has no default value. However, it won't influence matching.
E.g. Both `JobMetadataLock(type = XX)`, and `JobMetadataLock(type = YY)` match to my JobMetadataLockInterceptor.

**Reply by mkouba** [reply]
Yes, `@jakarta.enterprise.util.Nonbinding` is indeed the way to go. It means that the value is not used during interceptor resolution. In other words, the interceptor can define any value as it's ignored.

Now if you annotate the `serviceName()` with `@Nonbinding` you can still obtain the value declared on the bean (i.e. `FruiteResource`) from the `InvocationContext`. 

Unfortunately, there is no standardized way to obtain the bindings right now (it should be part of the jakarta interceptors 2.2). You can either use `io.quarkus.arc.ArcInvocationContext` instead of `jakarta.interceptor.InvocationContext` as the parameter in the `@AroundInvoke` method and then for example `ArcInvocationContext#findIterceptorBinding(Loggedclass)` or obtain the interceptor bindings from the context data, `(Set<Annotation>) InvocationContext#getContextData("io.quarkus.arc.interceptorBindings").`, and filter them manually.
